pep-9999.rst

```diff
@@ -0,0 +1,252 @@
+    
+variadic
+    A variadic type parameter is either co- or contravariant. A variadic
+    type (or variadic class) is a generic type with at least one variadic 
+    type parameter
+    
+
+Rationale
+=========
+
+In PEP 484 it states:
+
+    Covariance or contravariance is not a property of a type variable,
+    but a property of a generic class defined using this variable. 
+
+More specifically, variance is a property of the *type parameters* of a 
+generic class. 
+
+Currently, variadic type parameters are declared with type varianbles
+that are constructed with ``covariant=True`` or ``contravariant=True``, 
+e.g.:
+::
+
+    from typing import TypeVar, Generic, Protocol
+    
+    T_co = TypeVar("T_co", covariant=True)
+    T_contra = TypeVar("T_contra", contravariant=True)
+    
+    class SimpleFunction(Protocol[T_contra, T_co]):
+        def __call__(self, x: T_contra) -> T_co: ...
+
+But type variables can also be used outside of the scope of generic 
+types, where the concept of variance does not apply, e.g.:
+::
+
+    def first(things: Iterable[T]) -> T:
+        return next(iter(things))
+
+
+This example is valid, but using ``T_co`` or ``T_contra`` instead of 
+``T`` here does not make sense, and type checkers do not allow is.
+
+In ``SimpleFunction``, ``T_co`` or ``T_contra`` are used to annotate
+its method as well. This is required because the variance is associated 
+with the individual type variables. But the method signature is unrelated 
+to the variance of the type parameters of its generic type, 
+``SimpleFunction`` in this example. 
+
+Because variance is restricted to the type parameters of generic types, 
+it makes more sense to specify the variance of type parameters directly
+where the generic type is defined, removing the need to specify it on 
+the type variables themselves. 
+
+
+Proposal
+========
+
+The variance of the type parameters of a generic class can be specified
+with the ``+`` and ``-`` prefix operators on their respective type
+variables within the type parameter list of ``typing.Generic``, 
+``typing.Protocol``, or the generic base class. Instead of
+::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+    class SupportsInvertOld(typing.Protocol[T_co]):
+        def __invert__(self) -> T_co: ...
+
+    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
+        def __le__(self, other: T_contra): ...
+
+
+The new syntax uses the ``+`` and ``-`` prefix operators to specify
+variadic type parameters in the same place where the generic type is
+declared:
+::
+
+    T = TypeVar("T")
+    
+    # covariant
+    class SupportsInvert(typing.Protocol[+T]):
+        def __invert__(self) -> T: ...
+
+    # contravariant
+    class SupportsPartialOrder(typing.Protocol[-T]):
+        def __le__(self, other: T): ...
+
+This syntax is inspired by Scala programming language [1]_.
+
+
+
+Specification
```

Comment:
How should + and - work for a type variable with constraints? e.g.

```
T = TypeVar("T", str, bytes)
```

Reply:
I suppose that in this case it could make sense to use`T_A` for covariance, and `T_B` for contravariance.

```python
class A: ...
class B(A): ...

T_A = TypeVar("T_A", str, A)
T_B = TypeVar("T_B", str, B)
```

Reply:
I'm sorry, I'm not sure what you're trying to say with this example.

My understanding of constraints is that `T` can only be substituted for `str` or `bytes` and not subtypes of either. This means that variance probably shouldn't apply to such type variables?

Reply:
I don't know what variance would mean on a TypeVar with value restriction, so I'd be inclined to disallow it. But both mypy and pyright allow setting covariant=True on such a TypeVar, so maybe it means something I can't think of.

Regardless, it should be specified in the PEP. (And I also don't understand what the `T_A` example has to do with this topic.)

Reply:
The example does not make much sense in hindsight. I was trying to illustrate how a typevars are covariant in their upper bounds. Apparently this is used in Java on wildcard types ([wikipedia](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)#Use-site_variance_annotations_.28wildcards.29)), and in Scala, it is used when the type bound is another type parameter ([Scala docs](https://docs.scala-lang.org/tour/lower-type-bounds.html)).

In python, this would look something like `T = TypeVar('T', bound=+B)`, with `B` another typevar. But since this is not allowed at the moment, I agree that we shouldn't allow any constraints or bounds on variant typevars.